### PR TITLE
Add Notifier to Vinyl core

### PR DESCRIPTION
**Pull request by dscleaver**
Add the Notifier trait and support for loading Notifiers from config

Changes in this pull request:
- Add Notifier trait
- Add Notification wrapper class
- Add NotifierConfig class
- Add NotifierProvider trait
- Add NotifierLoader object
- Add AllNotifiers wrapper for loaded Notifier instances

### CODE REVIEW

File: modules/core/src/main/scala/vinyldns/core/notifier/NotifierLoader.scala [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+package vinyldns.core.notifier
+import vinyldns.core.domain.membership.UserRepository
+import cats.effect.IO
+import cats.implicits._
+
+object NotifierLoader {
+
+  def loadAll(configs: List[NotifierConfig], userRepository: UserRepository): IO[Notifier] =
```
Comment (rebstar6) [inline modules/core/src/main/scala/vinyldns/core/notifier/NotifierLoader.scala]: even in the case of a single (or no) notifiers, we are saying we will wrap in an AllNotifiers. I think that's fine, though i would then probably just have this return IO[AllNotifiers] and AllNotifiers doesnt need to extend the Notifier trait

also of note, if we didnt have this precedent, id say just have the load be object:apply for AllNotifiers. i could be convinced both to leave it here or put it there.

Comment (dscleaver) [reply modules/core/src/main/scala/vinyldns/core/notifier/NotifierLoader.scala]: hmmm.. I think I was comfortable hiding the specific implementation details. I can see your point though that contractually, `AllNotifiers` would then capture the "this should never fail" requirement

File: modules/core/src/main/scala/vinyldns/core/notifier/AllNotifiers.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+import cats.effect.{ContextShift, IO}
+import cats.implicits._
+import scala.concurrent.ExecutionContext
+import org.slf4j.LoggerFactory
+
+final case class AllNotifiers(notifiers: List[Notifier]) extends Notifier {
+
+  private val logger = LoggerFactory.getLogger("AllNotifiers")
+  private implicit val cs: ContextShift[IO] = IO.contextShift(ExecutionContext.global)
```
Comment (pauljamescleary) [inline modules/core/src/main/scala/vinyldns/core/notifier/AllNotifiers.scala]: probably makes sense to have the context shift passed in implicitly, either on the notify or more likely the case class itself.

File: modules/core/src/test/scala/vinyldns/core/notifier/NotifierLoaderSpec.scala [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+import cats.effect.IO
+import org.mockito.Mockito._
+
+import scala.collection.JavaConverters._
+
+object Mocks extends MockitoSugar {
+  val mockUserRepository: UserRepository = mock[UserRepository]
+}
+
+object MockNotifierProvider extends MockitoSugar {
```
Comment (pauljamescleary) [inline modules/core/src/test/scala/vinyldns/core/notifier/NotifierLoaderSpec.scala]: why are these separate objects?

File: modules/core/src/test/scala/vinyldns/core/notifier/NotifierLoaderSpec.scala [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    "return a notifier for valid config of one notifier" in {
+      val notifier = NotifierLoader.loadAll(List(goodConfig), mockUserRepository).unsafeRunSync()
+
+      notifier shouldNot be(null)
+
+      val notification = Notification(3)
+
+      reset(mockNotifier)
```
Comment (pauljamescleary) [inline modules/core/src/test/scala/vinyldns/core/notifier/NotifierLoaderSpec.scala]: If this is to reset mocks between tests, use the `BeforeEach` from ScalaTest and do a `reset` in the `def beforeEach`

File: modules/core/src/main/scala/vinyldns/core/notifier/AllNotifiers.scala (hunk 2 above)
Comment (dscleaver) [reply modules/core/src/main/scala/vinyldns/core/notifier/AllNotifiers.scala]: I was thinking this originally. I think it makes sense to pass it through load... grab it if you need it.

File: modules/core/src/test/scala/vinyldns/core/notifier/NotifierLoaderSpec.scala [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  import Mocks._
+  import MockNotifierProvider._
+
+  "loadAll" should {
+
+    "return some notifier with no configs" in {
+
+      val notifier = NotifierLoader.loadAll(List.empty, mockUserRepository).unsafeRunSync()
+
+      notifier shouldNot be(null)
```
Comment (mitruly) [inline modules/core/src/test/scala/vinyldns/core/notifier/NotifierLoaderSpec.scala]: `notifier shouldBe a[Notifier]` seems better than non-`null` check

File: modules/core/src/test/scala/vinyldns/core/notifier/AllNotifiersSpec.scala [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+  "notifier" should {
+    "notify all contained notifiers" in {
+
+      val notifier = AllNotifiers(mockNotifiers)
+
+      val notification = Notification("anything")
+
+      mockNotifiers.foreach { mock =>
+        reset(mock)
```
Comment (nimaeskandary) [inline modules/core/src/test/scala/vinyldns/core/notifier/AllNotifiersSpec.scala]: An alternative to sharing a single mock object and calling reset is having a fixture that provides a new mock for each test

```
trait Fixture {
    val mockNotifiers = List.fill(3)(mock[Notifier])
}
"notifier" should {
    "notify all contained notifiers" in new Fixture {...}
    "suppress errors from notifiers" in new Fixture{...}
}
```

File: modules/core/src/test/scala/vinyldns/core/notifier/NotifierLoaderSpec.scala (hunk 4 above)
Comment (nimaeskandary) [reply modules/core/src/test/scala/vinyldns/core/notifier/NotifierLoaderSpec.scala]: Made a comment above about fixtures as an alternative to beforeEach

File: modules/core/src/test/scala/vinyldns/core/notifier/NotifierLoaderSpec.scala (hunk 3 above)
Comment (dscleaver) [reply modules/core/src/test/scala/vinyldns/core/notifier/NotifierLoaderSpec.scala]: I wasn't thinking... changed it.

File: modules/core/src/test/scala/vinyldns/core/notifier/NotifierLoaderSpec.scala [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+  override def beforeEach: Unit =
+    reset(mockNotifier)
+
+  "loadAll" should {
+
+    "return some notifier with no configs" in {
+
+      val notifier = NotifierLoader.loadAll(List.empty, mockUserRepository).unsafeRunSync()
+
+      notifier shouldBe a[AllNotifiers]
```
Comment (mitruly) [inline modules/core/src/test/scala/vinyldns/core/notifier/NotifierLoaderSpec.scala]: Can use `an[AllNotifiers]` here.

**Pull request merged**